### ISSUE TITLE
Test failure: JIT/Directed/tailcall/mutual_recursion/mutual_recursion.cmd

### ISSUE THREAD

**Issue by v-wenyuxu** [dotnet/runtime#106603]
**Failed in:** [runtime-coreclr gcstress0x3-gcstress0xc 20240818.1](https://dev.azure.com/dnceng-public/public/_build/results?buildId=779491&view=ms.vss-test-web.build-test-results-tab&runId=19977360&resultId=115663&paneView=debug)

**Failed tests:**
```
coreclr linux arm Checked gcstress0x3 @ (Debian.12.Arm32.Open)Ubuntu.2004.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:debian-12-helix-arm32v7
    - JIT/Directed/tailcall/mutual_recursion/mutual_recursion.cmd
coreclr linux arm64 Checked gcstress0x3 @ (Ubuntu.2004.Arm64.Open)Ubuntu.2004.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-20.04-helix-arm64v8
    - JIT/Directed/tailcall/mutual_recursion/mutual_recursion.cmd
```

**Error message:**
```
 
cmdLine:/root/helix/work/workitem/e/JIT/Directed/Directed_3/../tailcall/mutual_recursion/mutual_recursion.sh Timed Out (timeout in milliseconds: 3600000 from variable __TestTimeout, start: 8/18/2024 3:15:56 PM, end: 8/18/2024 4:15:56 PM)

Return code:      -100
Raw output file:      /root/helix/work/workitem/uploads/tailcall/mutual_recursion/output.txt
Raw output:
BEGIN EXECUTION
/root/helix/work/correlation/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false -p System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true mutual_recursion.dll ''
[FirstMethod] - 158.499500ms
[SecondMethod] - 22.000000ms
[ThirdMethod] - 29540.908900ms
[FourthMethod] - 193.637200ms
[FifthMethod] - 2980864.325900ms

cmdLine:/root/helix/work/workitem/e/JIT/Directed/Directed_3/../tailcall/mutual_recursion/mutual_recursion.sh Timed Out (timeout in milliseconds: 3600000 from variable __TestTimeout, start: 8/18/2024 3:15:56 PM, end: 8/18/2024 4:15:56 PM)
Printing stacktrace from '/home/helixbot/dotnetbuild/dumps/crashdump_260.dmp.crashreport.json'
Test Harness Exitcode is : -100
To run the test:
Set up CORE_ROOT and run.
> /root/helix/work/workitem/e/JIT/Directed/Directed_3/../tailcall/mutual_recursion/mutual_recursion.sh
```

**Stack trace:**
```
   at Xunit.Assert.True(Nullable`1 condition, String userMessage) in /_/src/Microsoft.DotNet.XUnitAssert/src/BooleanAsserts.cs:line 146
   at TestLibrary.OutOfProcessTest.RunOutOfProcessTest(String assemblyPath, String testPathPrefix)
   at Program.<<Main>$>g__TestExecutor241|0_242(StreamWriter tempLogSw, StreamWriter statsCsvSw, <>c__DisplayClass0_0&)
```

**Comment by v-wenyuxu** [comment]
**Failed in:** [runtime-coreclr gcstress0x3-gcstress0xc 20240825.1](https://dev.azure.com/dnceng-public/public/_build/results?buildId=787819&view=ms.vss-test-web.build-test-results-tab&runId=20239174&resultId=115668&paneView=debug)

**Failed tests:**
```
coreclr linux arm Checked gcstress0x3 @ (Debian.12.Arm32.Open)Ubuntu.2004.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:debian-12-helix-arm32v7
    - JIT/Directed/tailcall/mutual_recursion/mutual_recursion.cmd
coreclr linux arm64 Checked gcstress0x3 @ (Ubuntu.2004.Arm64.Open)Ubuntu.2004.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-20.04-helix-arm64v8
    - JIT/Directed/tailcall/mutual_recursion/mutual_recursion.cmd
```

**Error message:**
```
 
cmdLine:/root/helix/work/workitem/e/JIT/Directed/Directed_3/../tailcall/mutual_recursion/mutual_recursion.sh Timed Out (timeout in milliseconds: 3600000 from variable __TestTimeout, start: 8/25/2024 2:51:00 PM, end: 8/25/2024 3:51:00 PM)

Return code:      -100
Raw output file:      /root/helix/work/workitem/uploads/tailcall/mutual_recursion/output.txt
Raw output:
BEGIN EXECUTION
/root/helix/work/correlation/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false -p System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true mutual_recursion.dll ''
[FirstMethod] - 161.387500ms
[SecondMethod] - 21.829100ms
[ThirdMethod] - 29195.583800ms
[FourthMethod] - 197.980600ms
[FifthMethod] - 3114073.315400ms

cmdLine:/root/helix/work/workitem/e/JIT/Directed/Directed_3/../tailcall/mutual_recursion/mutual_recursion.sh Timed Out (timeout in milliseconds: 3600000 from variable __TestTimeout, start: 8/25/2024 2:51:00 PM, end: 8/25/2024 3:51:00 PM)
Printing stacktrace from '/home/helixbot/dotnetbuild/dumps/crashdump_259.dmp.crashreport.json'
Test Harness Exitcode is : -100
To run the test:
Set up CORE_ROOT and run.
> /root/helix/work/workitem/e/JIT/Directed/Directed_3/../tailcall/mutual_recursion/mutual_recursion.sh
```

**Stack trace:**
```
   at Xunit.Assert.True(Nullable`1 condition, String userMessage) in /_/src/Microsoft.DotNet.XUnitAssert/src/BooleanAsserts.cs:line 146
   at TestLibrary.OutOfProcessTest.RunOutOfProcessTest(String assemblyPath, String testPathPrefix)
   at Program.<<Main>$>g__TestExecutor241|0_242(StreamWriter tempLogSw, StreamWriter statsCsvSw, <>c__DisplayClass0_0&)
```

**Comment by jakobbotsch** [comment]
I was able to bisect this to the range 7963a49d38f38dc4e6cfe32cd0f91677111649e4..f9c08462ba81a2aacc9fd85d321e12080c17b9ee. In that range we picked up the preview 7 SDK, which seems like it also included a new fsc version. I see significant IL codegen differences that I think are leading to the differences. Even without GC stress the preview7 version is much, much slower than the previous version. Will try to distill the problem a bit and open an issue on the F# repo.
